helper/keysutil/lock_manager.go

```diff
@@ -54,27 +54,85 @@ type PolicyRequest struct {
 }
 
 type LockManager struct {
-	useCache bool
-	// If caching is enabled, the map of name to in-memory policy cache
-	cache sync.Map
-
-	keyLocks []*locksutil.LockEntry
+	useCache  bool
+	cacheType CacheType
+	cache     Cache
+	keyLocks  []*locksutil.LockEntry
+	lock      sync.RWMutex
 }
 
 func NewLockManager(cacheDisabled bool) *LockManager {
 	lm := &LockManager{
-		useCache: !cacheDisabled,
-		keyLocks: locksutil.CreateLocks(),
+		useCache:  !cacheDisabled,
+		cacheType: NotInitialized,
+		keyLocks:  locksutil.CreateLocks(),
+		lock:      sync.RWMutex{},
 	}
+	lm.ConvertCacheToSyncmap()
 	return lm
 }
 
+func (lm *LockManager) ConvertCacheToLRU(size int) error {
+	lm.lock.Lock()
+	defer lm.lock.Unlock()
+	// update LockManager
+	newCache, err := NewTransitLRU(size)
+	if err != nil {
+		return err
+	}
+	lm.cacheType = LRU
+	lm.cache = newCache
+	return nil
+}
+
+func (lm *LockManager) ConvertCacheToSyncmap() {
+	lm.lock.Lock()
+	defer lm.lock.Unlock()
+	// update LockManager
+	lm.cacheType = SyncMap
+	lm.cache = NewTransitSyncMap()
+}
+
+func (lm *LockManager) GetCacheType() CacheType {
+	// Locked because lm.cache could change
+	lm.lock.RLock()
+	defer lm.lock.RUnlock()
+	return lm.cacheType
+}
+
+func (lm *LockManager) GetCacheSize() int {
+	// Locked because lm.cache could change
+	lm.lock.RLock()
+	defer lm.lock.RUnlock()
+	return lm.cache.Size()
+}
+
+func (lm *LockManager) SetCacheSize(size int) error {
+	switch {
+	case size < 0:
+		return errors.New("cache size must be greater or equal to zero")
+	case size == lm.GetCacheSize():
+		return fmt.Errorf("cache size is already set at: %d", size)
```

Comment:
Perhaps we just return nil here, seems like this doesn't need to be an error case?

Reply:
I debated on this and am happy to change it. Here is my reasoning that led me to making it an error case:
- nothing is actually `set` and I want to provide feedback to a user who might be trying to leverage SetCacheSize as a mechanism to clear cache. Returning nil will ultimately provide a 'Success' response to the user which opens some room for ambiguity as to whether the cache was just set now and cleared or had been set previously. The error message removes that ambiguity imo.
- it is a clean way to avoid writing to vault's storage [here](https://github.com/hashicorp/vault/pull/6225/files#diff-ce466f4b2b0b2027c1bfac932657007bR54) without needing to call b.lm.GetCacheSize() and make a comparison. I wanted to avoid that call so that the lock manager logic really is all in the lock manager and I also don't want to write to the storage entry when there has been no update to the lock manager.

Thoughts?

Reply:
I mentioned this because i believe when we call `initializeCache` in the Factory function we would hit this error case and the Factory would return an error. But i'll also comment on your other points:

> nothing is actually set and I want to provide feedback to a user who might be trying to leverage SetCacheSize as a mechanism to clear cache. Returning nil will ultimately provide a 'Success' response to the user which opens some room for ambiguity as to whether the cache was just set now and cleared or had been set previously. The error message removes that ambiguity imo.

Since SetCacheSize doesn't guarantee a cache clear (it's more of a side effect than an intent) I think it's fine to return a "Success" response. The intent of the function is to set the size to X and if the value is already at X it's just a no-op, but IMO is still a success. If we need a Cache Clear mechanism in the future we can add a separate function for it.

> it is a clean way to avoid writing to vault's storage here without needing to call b.lm.GetCacheSize() and make a comparison. I wanted to avoid that call so that the lock manager logic really is all in the lock manager and I also don't want to write to the storage entry when there has been no update to the lock manager.

This is a good point, but it is a common pattern in Vault to write to storage without first reading it to make sure we are not duplicating data (e.g. If you write the same value to k/v over and over we will persist it each time) I think it's fine to still update storage even if the value isn't changing. Though another option is to change the set function to return `bool, error` where the bool is indicating that the cache was updated.

Reply:
This is great context - thank you. Since there doesn't seem to be any concern with updating the storage even if the cache size value hasn't changed I think that is the cleanest solution. I'm inclined to avoid returning the bool following your reasoning that cache clearing is a side effect and not the intended purpose of setCache() (and that we can add something specific for clearing the cache if there is a need for it later).